### ISSUE TITLE
Read mnist code convolutional.py

### ISSUE THREAD

**Issue by cddesire** [cddesire/cddesire.github.io#36]

**Comment by cddesire** [comment]
**Strides** is a 4D array whose shape matches the data layout: **[image index, y, x, depth]**.
``` python
conv = tf.nn.conv2d(data,
                            conv1_weights,
                            strides=[1, 1, 1, 1],
                            padding='SAME')
```

**Comment by cddesire** [comment]
**Pattern**:  INPUT - CONV - RELU - POOL - FC
In more detail:

-  INPUT [32x32x3] will hold the raw pixel values of the image, in this case an image of width 32, height 32, and with three color channels R,G,B.
- CONV layer will compute the output of neurons that are connected to local regions in the input, each computing a dot product between their weights and a small region they are connected to in the input volume. This may result in volume such as [32x32x12] if we decided to use 12 filters.
- RELU layer will apply an elementwise activation function, such as the max(0,x) thresholding at zero. This leaves the size of the volume unchanged ([32x32x12]).
- POOL layer will perform a downsampling operation along the spatial dimensions (width, height), resulting in volume such as [16x16x12].
- FC (i.e. fully-connected) layer will compute the class scores, resulting in volume of size [1x1x10], where each of the 10 numbers correspond to a class score, such as among the 10 categories of CIFAR-10. As with ordinary Neural Networks and as the name implies, each neuron in this layer will be connected to all the numbers in the previous volume.

**Comment by cddesire** [comment]
### nn.sparse_softmax_cross_entropy_with_logits
稀疏表示的label与输出层计算出来结果做对比
`nn.sparse_softmax_cross_entropy_with_logits(logits,label,name=None)`
logits表示从最后一个隐藏层线性变换输出的结果。假设类别数目为10，那么对于每个样本这个logits应该是个10维的向量，且没有经过归一化，所有这个向量的元素和不为1，然后这个函数会先将logits进行softmax归一化，然后与label表示的onehot向量比较，计算交叉熵。 
这个函数执行了三步：
``` python
sm=nn.softmax(logits)
onehot=tf.sparse_to_dense(label，…)
nn.sparse_cross_entropy(sm,onehot)
```

**Comment by cddesire** [comment]
### tf.train.exponential_decay
指数下降调整步长
``` python
tf.train.exponential_decay(learning_rate, global_, decay_steps, decay_rate, staircase=True/False)
```
global_ : 当前批处理在数据集中的位置

**Comment by cddesire** [comment]
``` python
def model(data, train=False):
    """The Model definition."""
    # 2D convolution, with 'SAME' padding (i.e. the output feature map has
    # the same size as the input). Note that {strides} is a 4D array whose
    # shape matches the data layout: [image index, y, x, depth].
    conv = tf.nn.conv2d(data, conv1_weights, strides=[1, 1, 1, 1], padding='SAME')
    # Bias and rectified linear non-linearity.
    relu = tf.nn.relu(tf.nn.bias_add(conv, conv1_biases))
    # Max pooling. The kernel size spec {ksize} also follows the layout of
    # the data. Here we have a pooling window of 2, and a stride of 2.
    pool = tf.nn.max_pool(relu, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')
    conv = tf.nn.conv2d(pool, conv2_weights, strides=[1, 1, 1, 1], padding='SAME')
    relu = tf.nn.relu6(tf.nn.bias_add(conv, conv2_biases))
    pool = tf.nn.max_pool(relu, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')
    # Reshape the feature map cuboid into a 2D matrix to feed it to the
    # fully connected layers.
    pool_shape = pool.get_shape().as_list()
    reshape = tf.reshape(pool, [pool_shape[0], pool_shape[1] * pool_shape[2] * pool_shape[3]])
    # Fully connected layer. Note that the '+' operation automatically
    # broadcasts the biases.
    hidden = tf.nn.relu(tf.matmul(reshape, fc1_weights) + fc1_biases)
    # Add a 50% dropout during training only. Dropout also scales
    # activations such that no rescaling is needed at evaluation time.
    if train:
        hidden = tf.nn.dropout(hidden, 0.5, seed=SEED)

    output = tf.matmul(hidden, fc2_weights) + fc2_biases
    return output

```

**Comment by abigailhb** [comment]
＃ perfect
